crates/extensions/tedge-p11-server/src/connection.rs

```diff
@@ -0,0 +1,83 @@
+//! A connection between tedge-p11-server and client that provides a way to send and receive frames.
+//!
+//! Because connecting to the UNIX socket is very cheap, we can use a simple approach of only
+//! sending one request/response per connection.
+
+use std::io::Read;
+use std::io::Write;
+use std::net::Shutdown;
+use std::os::unix::net::UnixStream;
+
+use serde::Deserialize;
+use serde::Serialize;
+use tracing::warn;
+
+use crate::service::ChooseSchemeRequest;
+use crate::service::ChooseSchemeResponse;
+use crate::service::SignRequest;
+use crate::service::SignResponse;
+
+pub struct Connection {
+    stream: UnixStream,
+}
+
+impl Connection {
+    pub fn new(stream: UnixStream) -> Self {
+        Self { stream }
+    }
+
+    /// Reads a frame and closes the reading half of the connection.
+    ///
+    /// NOTE: can only be called once
+    pub fn read_frame(&mut self) -> anyhow::Result<Frame1> {
+        let mut buf = Vec::new();
+        self.stream.read_to_end(&mut buf)?;
+
+        // will fail if not version 1
+        let frame: Frame = postcard::from_bytes(&buf)?;
+
+        let Frame::Version1(frame) = frame;
+
+        self.stream.shutdown(Shutdown::Read)?;
+        if let Err(err) = self.stream.shutdown(Shutdown::Read) {
+            warn!("Failed to shutdown connection reading half: {err:?}");
+        }
+
+        Ok(frame)
+    }
+
+    /// Writes a frame and closes the writing half of the connection.
+    ///
+    /// NOTE: can only be called once
+    pub fn write_frame(&mut self, frame: &Frame1) -> anyhow::Result<()> {
+        let frame = Frame::Version1(frame.clone());
+
+        let buf = postcard::to_allocvec(&frame)?;
+
+        self.stream.write_all(&buf)?;
+        self.stream.flush()?;
+        if let Err(err) = self.stream.shutdown(Shutdown::Write) {
+            warn!("Failed to shutdown connection writing half: {err:?}");
+        }
+
+        Ok(())
+    }
+}
+
+/// The actual frame that we serialize and send/receive.
+///
+/// This essentially just adds a version tag and should deal with cases when
+/// non-backwards compatible changes are added to new versions.
+#[non_exhaustive]
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+enum Frame {
+    Version1(Frame1),
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub enum Frame1 {
+    ChooseSchemeRequest(ChooseSchemeRequest),
```

Comment:
Adding an error case, is required to able to handle version in the future. See https://github.com/thin-edge/thin-edge.io/pull/3421#discussion_r2006018279 

```suggestion
pub enum Frame1 {
    ProtocolError,
    ChooseSchemeRequest(ChooseSchemeRequest),
```

Reply:
Fixed in 1a3990a2ef73622e318df14842e8ea8da4f77601. I've also taken a liberty to add an error message string.

I've also sorted out a bit the idea of versioning in my head and have a bit better understanding.
Basically, because the version tag is the first thing we send over after the connection is established, we can read only the first byte and decide everything from there. Currently with version 1, after reading the version byte, we expect:
1. Client to send over a `Frame1` value that is a request.
2. Client to close its sending half.
3. At this point server reads the request until EOF, closes reading half, sends a response which is a single `Frame1` that **is not a request** (can be an error) and closes sending half.
4. Client should read the `Frame1` response until EOF and close reading half.
5. At this point the connection is terminated on both ends.

But in the future, using a hypothetical version 2, we can in theory move away from that model and send multiple requests and responses over a single connection.
- old clients will keep working with new server, maintaining forwards compatibility (but this will probably be less common case)
- new clients, to work with older server, will need some version negotiation mechanism to know what's the server version to choose appropriate message to send, which achieves backwards compatibility. This can be, as you suggested, an error returned by the server when client uses a new request that server can't support, at which point client can open a new connection and use the correct version, or a server can send over its own version first, which the client can read and decide what type of request to send. Due to using a UNIX socket, bandwidth and latency are not a concern, so the solution should be chosen based on simplicity.

Reply:
> Fixed in [1a3990a](https://github.com/thin-edge/thin-edge.io/commit/1a3990a2ef73622e318df14842e8ea8da4f77601). I've also taken a liberty to add an error message string.

Perfect!

> I've also sorted out a bit the idea of versioning in my head and have a bit better understanding. Basically, because the version tag is the first thing we send over after the connection is established, we can read only the first byte and decide everything from there. Currently with version 1, after reading the version byte, we expect:

I would simply let `serde` figure out which version is the received frame. Responding with the same version or returning a protocol error if the frame cannot be deserialized - forcing the peer to downgrade its protocol.

>  But in the future, using a hypothetical version 2, we can in theory move away from that model and send multiple requests and responses over a single connection.

Possibly.